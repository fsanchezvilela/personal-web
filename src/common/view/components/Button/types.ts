import {  ReactNode } from "react"

export type ButtonStyleType = 'primary' | 'secondary' | 'tertiary' | 'disabled'

export type ButtonPropsType = {
  children: ReactNode
  buttonStyle?: ButtonStyleType,
  href?: string;
}
