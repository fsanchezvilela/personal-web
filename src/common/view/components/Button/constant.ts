export const ButtonStyleTwClassMapping = {
  'primary': "rounded border transition bg-transparent text-keppel  hover:bg-keppel-tint px-3 py-2",
  'secondary': "rounded border transition px-3 py-2 hover:bg-dark-jungle-green hover:culturedr-sand text-silver-sand" ,
  'tertiary': " rounded border-dark-jungle-green transition px-3 py-2 hover:text-silver-sand text-keppel",
  'disabled': 'cursor-not-allowed rounded border transition px-3 py-2 bg-silver-sand text-roman-silver'
}