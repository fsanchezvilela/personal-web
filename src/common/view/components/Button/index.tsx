import { ButtonHTMLAttributes, DetailedHTMLProps, FC } from "react";
import { ButtonStyleTwClassMapping } from "./constant";
import { ButtonPropsType } from "./types";

const Button: FC<ButtonPropsType & DetailedHTMLProps<ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement>> = ({ children, buttonStyle = 'primary', href = '', ...props }) => {

  const RenderButton = () => {
    const DisableButton = <button disabled className={ButtonStyleTwClassMapping['disabled']}>
      {children}
    </button>

    if (props.disabled === true) { return DisableButton }

    return <button {...props} className={ButtonStyleTwClassMapping[buttonStyle]}>{children}</button>
  }

  const RenderAnchor = () => <a className={ButtonStyleTwClassMapping[buttonStyle]} href={href}>{children}</a>

  return (
    href ?
      RenderAnchor() :
      RenderButton()
    )
}


Button.displayName = 'Button';
export default Button
