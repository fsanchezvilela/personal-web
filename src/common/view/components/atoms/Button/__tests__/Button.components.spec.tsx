import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import Button from '..'
import { ButtonStyleTwClassMapping } from '../constant'

describe('Button Component Test', () => {

  const { primary, secondary, tertiary, disabled } = ButtonStyleTwClassMapping;

  it('Renders a Primary Button Default', () => {
    const childrenRender = `click me! default button`
    const { asFragment } = render(<Button>{childrenRender}</Button>)
    // See if render children correct
    expect(asFragment().textContent).toMatch(childrenRender)
    // See if have primary button styles 
    expect(asFragment().querySelector('button')?.className).toMatch(primary)

    expect(asFragment()).toMatchSnapshot();
  })

  it('Renders a Primary Button with props', () => {
    const childrenRender = `click me! primary button`
    const { asFragment } = render(<Button buttonStyle='primary'>{childrenRender}</Button>)
    // See if render children correct
    expect(asFragment().textContent).toMatch(childrenRender)
    // See if have primary button styles 
    expect(asFragment().querySelector('button')?.className).toMatch(primary)

    expect(asFragment()).toMatchSnapshot();
  })

  it('Renders a Secondary Button', () => {
    const childrenRender = `click me! secondary button`
    const { asFragment } = render(<Button buttonStyle='secondary'>{childrenRender}</Button>)
    // See if render children correct
    expect(asFragment().textContent).toMatch(childrenRender)
    // See if have secondary button styles 
    expect(asFragment().querySelector('button')?.className).toMatch(secondary)

    expect(asFragment()).toMatchSnapshot();
  })

  it('Renders a Tertiary Button', () => {
    const childrenRender = `click me! tertiary button`
    const { asFragment } = render(<Button buttonStyle='tertiary'>{childrenRender}</Button>)
    // See if render children correct
    expect(asFragment().textContent).toMatch(childrenRender)
    // See if have tertiary button styles 
    expect(asFragment().querySelector('button')?.className).toMatch(tertiary)

    expect(asFragment()).toMatchSnapshot();
  })

  it('Renders a Disabled Button', () => {
    const childrenRender = `disabled button`
    const { asFragment } = render(<Button disabled>{childrenRender}</Button>)
    // See if render children correct
    expect(asFragment().textContent).toMatch(childrenRender)
    // See if have disabled button styles 
    expect(asFragment().querySelector('button')?.className).toMatch(disabled)
    expect(asFragment()).toMatchSnapshot();
  })

})