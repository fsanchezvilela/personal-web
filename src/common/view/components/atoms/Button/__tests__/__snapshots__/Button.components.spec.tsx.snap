// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component Test Renders a Disabled Button 1`] = `
<DocumentFragment>
  <button
    class="cursor-not-allowed rounded border transition px-3 py-2 bg-silver-sand text-roman-silver"
    disabled=""
  >
    disabled button
  </button>
</DocumentFragment>
`;

exports[`Button Component Test Renders a Primary Button Default 1`] = `
<DocumentFragment>
  <button
    class="rounded border transition bg-transparent text-keppel  hover:bg-keppel-tint px-3 py-2"
  >
    click me! default button
  </button>
</DocumentFragment>
`;

exports[`Button Component Test Renders a Primary Button with props 1`] = `
<DocumentFragment>
  <button
    class="rounded border transition bg-transparent text-keppel  hover:bg-keppel-tint px-3 py-2"
  >
    click me! primary button
  </button>
</DocumentFragment>
`;

exports[`Button Component Test Renders a Secondary Button 1`] = `
<DocumentFragment>
  <button
    class="rounded border transition px-3 py-2 hover:bg-dark-jungle-green hover:culturedr-sand text-silver-sand"
  >
    click me! secondary button
  </button>
</DocumentFragment>
`;

exports[`Button Component Test Renders a Tertiary Button 1`] = `
<DocumentFragment>
  <button
    class=" rounded border-dark-jungle-green transition px-3 py-2 hover:text-silver-sand text-keppel"
  >
    click me! tertiary button
  </button>
</DocumentFragment>
`;
